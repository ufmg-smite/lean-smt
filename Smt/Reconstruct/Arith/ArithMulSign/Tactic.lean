/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Tomaz Gomes Mascarenhas
-/

import Lean

import Mathlib.Data.Nat.Parity
import Mathlib.Data.Real.Basic

import Smt.Reconstruct.Arith.ArithMulSign.Lemmas
import Smt.Reconstruct.Util

namespace Smt.Reconstruct.Arith

open Lean Elab Tactic Meta Expr

inductive Pol where
 | Neg : Pol -- 0
 | NZ  : Pol -- 1
 | Pos : Pol -- 2
 deriving BEq

def intLOR := mkApp2 (.const ``LinearOrderedCommRing.toLinearOrderedRing [.zero])
                     (.const ``Int []) (.const ``Int.linearOrderedCommRing [])

def RealLOR := Expr.const ``Real.instLinearOrderedRingReal []


def traceMulSign (r : Except Exception Unit) : MetaM MessageData :=
  return match r with
  | .ok _ => m!"{checkEmoji}"
  | _     => m!"{bombEmoji}"

def mulSign (mv : MVarId) (xs : Array (Expr × Fin 3 × Nat)) : MetaM Unit := withTraceNode `smt.reconstruct.mulSign traceMulSign do
  mv.assign (← go true xs.toList (mkConst `empty) (mkConst `empty))
where
  go (first : Bool) (xs : (List (Expr × Fin 3 × Nat))) (prodSignPf : Expr) (prod : Expr) : MetaM Expr :=
  match xs with
  | [] => return prodSignPf
  | (expr, pol, exp) :: t => do
    let exprType ← inferType expr
    let exprIsInt ←
      match exprType with
      | .const `Rat .. => pure false
      | .const `Int .. => pure true
      | _ => throwError "[arithMulSign]: unexpected type for expression"
    let lorInst := if exprIsInt then intLOR else RealLOR
    let zeroI := mkApp (mkConst ``Int.ofNat) (mkNatLit 0)
    let zeroR := mkApp (mkConst ``Rat.ofInt) zeroI
    -- zero with the same type as the current argument
    let currZero := if exprIsInt then zeroI else zeroR
    let bindedType: Expr ← match pol with
      | 0 => mkAppM ``LT.lt #[expr, currZero]
      | 1 => mkAppM ``Ne    #[expr, currZero]
      | 2 => mkAppM ``GT.gt #[expr, currZero]
    let expParityPf ← do
      if exp % 2 == 0 then
        let rflExpr ← mkAppOptM ``rfl #[(mkConst ``Nat), (mkNatLit 0)]
        let iffExp := mkApp (mkConst ``Nat.even_iff) (mkNatLit exp)
        mkAppM ``Iff.mpr #[iffExp, rflExpr]
      else
        let rflExpr ← mkAppOptM ``rfl #[(mkConst ``Nat), (mkNatLit 1)]
        let iffExp := mkApp (mkConst ``Nat.odd_iff) (mkNatLit exp)
        mkAppM ``Iff.mpr #[iffExp, rflExpr]
    withLocalDeclD (← mkFreshId) bindedType $ fun bv => do
      let exprPowSignPf ←
        if exp == 1 then
          pure bv
        else
          match pol with
          | 1 =>
            pure $ mkApp6 (mkConst ``nonZeroEvenPow) exprType lorInst (mkNatLit exp) expr bv expParityPf
          | 2 =>
            pure $ mkApp5 (mkConst ``powPos) exprType lorInst (mkNatLit exp) expr bv
          | 0 =>
            if exp % 2 == 0 then
              mkAppM ``powNegEven #[bv, expParityPf]
            else
              mkAppM ``powNegOdd #[bv, expParityPf]
      let exprPow ←
        if exp == 1 then
          pure expr
        else mkAppM ``Pow.pow #[expr, mkNatLit exp]
      let prodType ←
        if first then
          pure exprType
        else inferType prod
      let prodIsInt ←
        match prodType with
        | .const `Int .. => pure true
        | .const `Rat .. => pure false
        | _ => throwError "[arithMulSign]: unexpected type for accumulated product"
      let prodSignPfType ←
        if first then
          inferType exprPowSignPf
        else inferType prodSignPf
      let prodPos := (← getOp prodSignPfType) == `GT.gt
      -- normalize types in case one is rat and the other is int
      let (exprPow', prod', exprPowSignPf', prodSignPf') :=
        match exprIsInt, prodIsInt with
        | false, true  =>
          let prodSignPf' :=
            if prodPos then
              mkApp2 (mkConst ``castPos) prod prodSignPf
            else mkApp2 (mkConst ``castNeg) prod prodSignPf
          (exprPow, mkApp (mkConst ``Rat.ofInt) prod, exprPowSignPf, prodSignPf')
        | true, false  =>
          let exprPowSignPf' :=
            if pol == 2 || pol == 1 || exp % 2 == 0 then
              mkApp2 (mkConst ``castPos) exprPow exprPowSignPf
            else mkApp2 (mkConst ``castNeg) exprPow exprPowSignPf
          (mkApp (mkConst ``Rat.ofInt) exprPow, prod, exprPowSignPf', prodSignPf)
        | _, _   => (exprPow, prod, exprPowSignPf, prodSignPf)
      let answer ←
        if first then pure exprPowSignPf
        else
          if pol == 2 || pol == 1 || exp % 2 == 0 then
            if prodPos then
              mkAppOptM ``combineSigns₁ #[none, none, exprPow', prod', exprPowSignPf', prodSignPf']
            else mkAppOptM ``combineSigns₂ #[none, none, exprPow', prod', exprPowSignPf', prodSignPf']
          else
            if prodPos then
              mkAppOptM ``combineSigns₃ #[none, none, exprPow', prod', exprPowSignPf', prodSignPf']
            else mkAppOptM ``combineSigns₄ #[none, none, exprPow', prod', exprPowSignPf', prodSignPf']
      let prod' ←
        if first then
          pure exprPow
        else mkAppM ``Mul.mul #[prod', exprPow']
      let rc ← go false t answer prod'
      mkLambdaFVars #[bv] rc

namespace Tactic

--              var names/neg or pos/ exponents
-- arithMulSign [a, b, c] [1, -1, 1], [2, 3, 1]
syntax (name := arithMulSign) "arithMulSign" ("[" term,* "]")? "," ("[" term,* "]") "," ("[" term,* "]")? : tactic

def parseArithMulSign : Syntax → TacticM (List (Expr × Pol × Nat))
  | `(tactic| arithMulSign [ $[$ns],* ], [ $[$bs],* ], [ $[$es],* ]) => do
    let exprs: List Expr ←
      ns.toList.mapM (fun h: Term => elabTerm h none)
    let bools: List Pol ←
      bs.toList.mapM (fun h: Term => do
                      let hExpr ← elabTerm h none
                      match getNatLit? hExpr with
                      | none   => pure Pol.Neg
                      | some 0 => pure Pol.NZ
                      | some 1 => pure Pol.Pos
                      | _      => throwError "[arithMulSign]: invalid polarity"
                    )
    let exps : List Nat ← es.toList.mapM stxToNat
    pure (exprs.zip (bools.zip exps))
  | _ => throwError "[ArithMulSign]: wrong syntax"

@[tactic arithMulSign] def evalArithMulSign : Tactic := fun stx => do
  let mvar ← getMainGoal
  mvar.withContext do
    trace[smt.profile.reconstruct] m!"[arithMulSign] start time: {← IO.monoNanosNow}ns"
    let data ← parseArithMulSign stx
    let answer ← go true data (mkConst `empty) (mkConst `empty)
    let answerType ← inferType answer
    let fname ← mkFreshId
    let (_, mvar') ← MVarId.intro1P $ ← mvar.assert fname answerType answer
    replaceMainGoal [mvar']
    evalTactic (← `(tactic| exact $(mkIdent fname)))
    trace[smt.profile.reconstruct] m!"[arithMulSign] end time: {← IO.monoNanosNow}ns"
where
-- acc is a partial proof corresponding to the sign of the prefix of the multiplication
  go (first : Bool) (xs : (List (Expr × Pol × Nat))) (prodSignPf : Expr) (prod : Expr) : MetaM Expr :=
    match xs with
    | [] => return prodSignPf
    | (expr, pol, exp) :: t => do
      let exprType ← inferType expr
      let exprIsInt ←
        match exprType with
        | .const `Real .. => pure false
        | .const `Int .. => pure true
        | _ => throwError "[arithMulSign]: unexpected type for expression"
      let lorInst := if exprIsInt then intLOR else RealLOR
      let zeroI := mkApp (mkConst ``Int.ofNat) (mkNatLit 0)
      let zeroR ← mkAppOptM' (.const ``OfNat.ofNat [.zero]) #[mkConst ``Real, (mkNatLit 0), none]
      -- zero with the same type as the current argument
      let currZero := if exprIsInt then zeroI else zeroR
      let bindedType: Expr ←
        match pol with
        | Pol.Neg => mkAppM ``LT.lt #[expr, currZero]
        | Pol.NZ  => mkAppM ``Ne    #[expr, currZero]
        | Pol.Pos => mkAppM ``GT.gt #[expr, currZero]
      let expParityPf ← do
        if exp % 2 == 0 then
          let rflExpr ← mkAppOptM ``rfl #[(mkConst ``Nat), (mkNatLit 0)]
          let iffExp := mkApp (mkConst ``Nat.even_iff) (mkNatLit exp)
          mkAppM ``Iff.mpr #[iffExp, rflExpr]
        else
          let rflExpr ← mkAppOptM ``rfl #[(mkConst ``Nat), (mkNatLit 1)]
          let iffExp := mkApp (mkConst ``Nat.odd_iff) (mkNatLit exp)
          mkAppM ``Iff.mpr #[iffExp, rflExpr]
      withLocalDeclD (← mkFreshId) bindedType $ fun bv => do
        let exprPowSignPf ←
          if exp == 1 then
            pure bv
          else
            match pol with
            | Pol.NZ =>
              pure $ mkApp6 (mkConst ``nonZeroEvenPow) exprType lorInst (mkNatLit exp) expr bv expParityPf
            | Pol.Pos =>
              pure $ mkApp5 (mkConst ``powPos) exprType lorInst (mkNatLit exp) expr bv
            | Pol.Neg =>
              if exp % 2 == 0 then
                mkAppM ``powNegEven #[bv, expParityPf]
              else
                mkAppM ``powNegOdd #[bv, expParityPf]
        let exprPow ←
          if exp == 1 then
            pure expr
          else mkAppM ``Pow.pow #[expr, mkNatLit exp]
        let prodType ←
          if first then
            pure exprType
          else inferType prod
        let prodIsInt ←
          match prodType with
          | .const `Int .. => pure true
          | .const `Real .. => pure false
          | _ => throwError "[arithMulSign]: unexpected type for accumulated product"
        let prodSignPfType ←
          if first then
            inferType exprPowSignPf
          else inferType prodSignPf
        let prodPos := (← getOp prodSignPfType) == `GT.gt
        -- normalize types in case one is real and the other is int
        let (exprPow', prod', exprPowSignPf', prodSignPf') :=
          match exprIsInt, prodIsInt with
          | false, true  =>
            let prodSignPf' :=
              if prodPos then
                mkApp2 (mkConst ``castPos) prod prodSignPf
              else mkApp2 (mkConst ``castNeg) prod prodSignPf
            (exprPow, mkApp (mkConst ``zToR) prod, exprPowSignPf, prodSignPf')
          | true, false  =>
            let exprPowSignPf' :=
              if pol == Pol.Pos || pol == Pol.NZ || exp % 2 == 0 then
                mkApp2 (mkConst ``castPos) exprPow exprPowSignPf
              else mkApp2 (mkConst ``castNeg) exprPow exprPowSignPf
            (mkApp (mkConst ``zToR) exprPow, prod, exprPowSignPf', prodSignPf)
          | _, _   => (exprPow, prod, exprPowSignPf, prodSignPf)
        let answer ←
          if first then pure exprPowSignPf
          else
            if pol == Pol.Pos || pol == Pol.NZ || exp % 2 == 0 then
              if prodPos then
                mkAppOptM ``combineSigns₁ #[none, none, exprPow', prod', exprPowSignPf', prodSignPf']
              else mkAppOptM ``combineSigns₂ #[none, none, exprPow', prod', exprPowSignPf', prodSignPf']
            else
              if prodPos then
                mkAppOptM ``combineSigns₃ #[none, none, exprPow', prod', exprPowSignPf', prodSignPf']
              else mkAppOptM ``combineSigns₄ #[none, none, exprPow', prod', exprPowSignPf', prodSignPf']
        let prod' ←
          if first then
            pure exprPow
          else mkAppM ``Mul.mul #[prod', exprPow']
        let rc ← go false t answer prod'
        mkLambdaFVars #[bv] rc

end Smt.Reconstruct.Arith.Tactic
