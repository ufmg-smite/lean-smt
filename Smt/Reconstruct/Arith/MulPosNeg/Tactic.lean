/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Tomaz Gomes Mascarenhas
-/

import Lean

import Mathlib.Data.Real.Basic

import Smt.Reconstruct.Arith.MulPosNeg.Lemmas
import Smt.Reconstruct.Util

namespace Smt.Reconstruct.Arith

open Lean
open Elab Tactic Expr Meta

syntax (name := arithMulPos) "arithMulPos" ("[" term,* "]")? "," term : tactic
syntax (name := arithMulNeg) "arithMulNeg" ("[" term,* "]")? "," term : tactic

def parseArithMulAux : Array Term → Term → TacticM (Expr × Expr × Expr × Nat)
  | hs, i => do
    let li: List Expr ←
      hs.toList.mapM (fun h: Term => elabTerm h none)
    let i' ← stxToNat i
    match li with
    | [a, b, c] => return (a, b, c, i')
    | _ => throwError "[arithMul]: List must have 3 elements"

def parseArithMul : Syntax → TacticM (Expr × Expr × Expr × Nat)
  | `(tactic| arithMulPos [ $[$hs],* ], $i) => parseArithMulAux hs i
  | `(tactic| arithMulNeg [ $[$hs],* ], $i) => parseArithMulAux hs i
  | _ => throwError "[arithMul]: wrong usage"

def operators : List Name :=
  [``LT.lt, ``LE.le, ``GT.gt, ``GE.ge, ``Eq]

def castFsts : List Name :=
  [``castFstLT , ``castFstLE , ``castFstGT , ``castFstGE , ``castFstEQ]

def castSnds : List Name :=
  [``castSndLT , ``castSndLE , ``castSndGT , ``castSndGE , ``castSndEQ]

def arithMulMeta (va vb vc : Expr) (pos : Bool) (compId : Nat) (thms : List Name) :
    MetaM Expr := do
  let mut typeA ← inferType va
  let mut typeB ← inferType vb
  let mut va := va
  let mut vb := vb
  if typeA != typeB then
    if typeA == mkConst ``Int then
      va := mkApp (mkConst ``zToR) va
      typeA := mkConst ``Real
    else
      vb := mkApp (mkConst ``zToR) vb
      typeB := mkConst ``Real
  let typeC ← inferType vc
  let thmName ←
    if compId < 5 then
      pure (thms.get! compId)
    else throwError "[arithMul]: index too large"

  let zeroI := mkApp (mkConst ``Int.ofNat) (mkNatLit 0)
  let zeroR ← mkAppOptM' (.const ``OfNat.ofNat [.zero]) #[mkConst ``Real, (mkNatLit 0), none]
  let zeroC := if typeC == mkConst ``Int then zeroI else zeroR
  let premiseLeft ←
    if pos then mkAppM ``LT.lt #[zeroC, vc]
    else mkAppM ``LT.lt #[vc, zeroC]
  let operator ←
    if compId < 5 then
      pure $ operators.get! compId
    else throwError "[arithMul]: index too large"
  let premiseRight ← mkAppM operator #[va, vb]

  let premiseType :=
    mkApp2 (mkConst ``And) premiseLeft premiseRight

  match typeA, typeC with
    | const ``Int _, const ``Int _ =>
      mkAppM thmName #[]
    | const ``Int _, const ``Real _ =>
      withLocalDeclD (← mkFreshId) premiseType $ fun bv => do
        let e₁ ← mkAppM (castSnds.get! compId) #[bv]
        let e₂ ← mkAppM thmName #[e₁]
        mkLambdaFVars #[bv] e₂
    | const ``Real _, const ``Int _ =>
      withLocalDeclD (← mkFreshId) premiseType $ fun bv => do
        let e₁ ← mkAppM (castFsts.get! compId) #[bv]
        let e₂ ← mkAppM thmName #[e₁]
        mkLambdaFVars #[bv] e₂
    | const ``Real _, const ``Real _ =>
      mkAppM thmName #[]
    | _, _ => throwError "[arithMul]: unexpected variable type"

@[tactic arithMulPos] def evalArithMulPos : Tactic := fun stx => do
  trace[smt.profile.reconstruct] m!"[arithMulPos] start time: {← IO.monoNanosNow}ns"
  let (a, b, c, compId) ← parseArithMul stx
  let mvar ← getMainGoal
  let pf ← arithMulMeta a b c true compId
                [ ``arith_mul_pos_lt
                , ``arith_mul_pos_le
                , ``arith_mul_pos_gt
                , ``arith_mul_pos_ge
                , ``arith_mul_pos_eq
                ]
  let type ← inferType pf
  let fname ← mkFreshId
  let (_, mvar') ← MVarId.intro1P $ ← mvar.assert fname type pf
  replaceMainGoal [mvar']
  evalTactic (← `(tactic| exact $(mkIdent fname)))
  trace[smt.profile.reconstruct] m!"[arithMulPos] end time: {← IO.monoNanosNow}ns"

@[tactic arithMulNeg] def evalArithMulNeg : Tactic := fun stx => do
  trace[smt.profile.reconstruct] m!"[arithMulNeg] start time: {← IO.monoNanosNow}ns"
  let (a, b, c, compId) ← parseArithMul stx
  let mvar ← getMainGoal
  let pf ← arithMulMeta a b c false compId
                [ ``arith_mul_neg_lt
                , ``arith_mul_neg_le
                , ``arith_mul_neg_gt
                , ``arith_mul_neg_ge
                , ``arith_mul_neg_eq
                ]
  let type ← inferType pf
  let fname ← mkFreshId
  let (_, mvar') ← MVarId.intro1P $ ← mvar.assert fname type pf
  replaceMainGoal [mvar']
  evalTactic (← `(tactic| exact $(mkIdent fname)))
  trace[smt.profile.reconstruct] m!"[arithMulNeg] end time: {← IO.monoNanosNow}ns"

end Smt.Reconstruct.Arith
