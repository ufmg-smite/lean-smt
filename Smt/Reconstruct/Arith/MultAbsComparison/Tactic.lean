/-
Copyright (c) 2021-2025 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Tomaz Gomes Mascarenhas
-/

import Lean
import Qq
import Smt.Reconstruct.Arith.MultAbsComparison.Lemmas

namespace Smt.Reconstruct.Arith

open Lean Elab Tactic Meta Expr
open Qq

syntax (name := multAbsComparison) "multAbsComparison" ("[" term,* "]")? : tactic

def multAbsComparison₁ (mvar : MVarId) (hs : List Expr) (fname : Name) : MetaM MVarId := do
  let h₁ :: hs' ← pure hs | throwError "unreachable"
  go h₁ hs'
where
  go (acc : Expr) (es : List Expr) : MetaM MVarId :=
    match es with
    | [] => do
      let type ← inferType acc
      let (_, mvar') ← MVarId.intro1P $ ← mvar.assert fname type acc
      return mvar'
    | h :: hs => do
      let th ← inferType h
      let thm ← match th with
      | app (app (app (const ``GT.gt ..) _) _) _ => pure ``mult_abs_0
      | app (app (const ``And ..) _) _ => pure ``mult_abs_2
      | _ => throwError "[multAbsComparison]: unexpected hypothesis"
      let acc' ← mkAppM thm #[acc, h]
      go acc' hs

def multAbsComparison₂ (mvar : MVarId) (hs : List Expr) (fname : Name) : MetaM MVarId := do
  let h₁ :: hs' ← pure hs | throwError "unreachable"
  go h₁ hs'
where
  go (acc : Expr) (es : List Expr) : MetaM MVarId :=
    match es with
    | [] => do
      let type ← inferType acc
      let (_, mvar') ← MVarId.intro1P $ ← mvar.assert fname type acc
      return mvar'
    | h :: hs => do
      let acc' ← mkAppM ``mult_abs_eq #[acc, h]
      go acc' hs

def parseMultAbsComparison : Syntax → TacticM (List Expr)
  | `(tactic| multAbsComparison [$[$hs],*]) =>
    hs.toList.mapM (fun stx => elabTerm stx.raw none)
  | _ => throwError "[multAbsComparison]: expects a list of premisses"

@[tactic multAbsComparison] def evalMultAbsComparison : Tactic := fun stx => do
  let mvar ← Tactic.getMainGoal
  mvar.withContext do
    let h₁::hs ← parseMultAbsComparison stx | throwError "[multAbsComparison]: empty list of hypothesis"
    let t₁ ← inferType h₁
    let fname ← mkFreshId
    let mvar' ← match t₁ with
    | app (app (app (const ``GT.gt ..) _) _) _ => multAbsComparison₁ mvar (h₁ :: hs) fname
    | app (app (app (const ``Eq ..) _) _) _ => multAbsComparison₂ mvar (h₁ :: hs) fname
    | _ => throwError "[multAbsComparison]: unexpected input - first hypothesis is neither = or >"
    replaceMainGoal [mvar']
    evalTactic (← `(tactic| exact $(mkIdent fname)))

example (x1 y1 x2 y2 x3 y3 : Int) : abs x1 = abs y1 → abs x2 = abs y2 → abs x3 = abs y3 → abs (x1 * x2 * x3) = abs (y1 * y2 * y3) := by
  intros h1 h2 h3
  multAbsComparison [h1, h2, h3]
