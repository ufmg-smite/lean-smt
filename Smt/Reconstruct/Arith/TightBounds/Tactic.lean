/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Tomaz Gomes Mascarenhas
-/

import Smt.Reconstruct.Arith.TightBounds.Lemmas
import Smt.Reconstruct.Util

import Lean
import Mathlib.Algebra.Order.Floor

namespace Smt.Reconstruct.Arith

open Lean hiding Rat
open Meta Elab.Tactic Expr

def isIntLt : Expr → Bool
  | app (app (app (app _ (const ``Int ..)) ..) ..) .. => true
  | _ => false

def intTightMeta (mvar : MVarId) (h : Expr) (thmName outName : Name)
  : MetaM MVarId :=
  mvar.withContext do
    let t ← inferType h
    let arg ←
      if isIntLt t then
        mkAppM ``castLT #[h]
      else pure h
    let answer ← mkAppM thmName #[arg]
    let answerType ← inferType answer
    let (_, mvar') ← MVarId.intro1P $ ← mvar.assert outName answerType answer
    return mvar'

syntax (name := intTightUb) "intTightUb" term : tactic
@[tactic intTightUb] def evalIntTightUb : Tactic := fun stx =>
  withMainContext do
    trace[smt.debug] m!"[intTightUb] start time: {← IO.monoNanosNow}ns"
    let h ← elabTerm stx[1] none
    let fname ← mkFreshId
    let mvar ← getMainGoal
    let mvar' ← intTightMeta mvar h ``intTightUb' fname
    replaceMainGoal [mvar']
    evalTactic (← `(tactic| exact $(mkIdent fname)))
    trace[smt.debug] m!"[intTightUb] end time: {← IO.monoNanosNow}ns"

syntax (name := intTightLb) "intTightLb" term : tactic
@[tactic intTightLb] def evalIntTightLb : Tactic := fun stx =>
  withMainContext do
    trace[smt.debug] m!"[intTightLb] start time: {← IO.monoNanosNow}ns"
    let h ← elabTerm stx[1] none
    let fname ← mkFreshId
    let mvar ← getMainGoal
    let mvar' ← intTightMeta mvar h ``intTightLb' fname
    replaceMainGoal [mvar']
    evalTactic (← `(tactic| exact $(mkIdent fname)))
    trace[smt.debug] m!"[intTightLb] end time: {← IO.monoNanosNow}ns"

end Smt.Reconstruct.Arith
