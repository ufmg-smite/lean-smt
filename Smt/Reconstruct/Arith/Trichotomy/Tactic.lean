/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Tomaz Gomes Mascarenhas
-/

import Lean

-- we must import those to have visible instances of LinearOrder of
-- Nat, Int and Rat
import Mathlib.Data.Rat.Order
import Mathlib.Init.Data.Int.Order
import Mathlib.Init.Data.Nat.Lemmas

import Smt.Reconstruct.Arith.Trichotomy.Lemmas

namespace Smt.Reconstruct.Arith

open Lean hiding Rat
open Meta Elab.Tactic Expr

def trichotomyMeta (mvar : MVarId) (h₁ h₂ : Expr) (name : Name) : MetaM MVarId :=
  mvar.withContext do
    let t₁ ← inferType h₁
    let t₂ ← inferType h₂
    let thmName : Name ←
      match ← getOp t₁, ← getOp t₂ with
      | ``LT.lt , ``Eq    => pure ``trichotomy₁
      | ``Eq    , ``LT.lt => pure ``trichotomy₂
      | ``LT.lt , ``GT.gt => pure ``trichotomy₃
      | ``GT.gt , ``LT.lt => pure ``trichotomy₄
      | ``Eq    , ``GT.gt => pure ``trichotomy₅
      | ``GT.gt , ``Eq    => pure ``trichotomy₆
      | _      , _      => throwError "[trichotomy] invalid operation"
    let answer ← mkAppM thmName #[h₁, h₂]
    let answerType ← inferType answer
    let (_, mvar') ← MVarId.intro1P $ ← mvar.assert name answerType answer
    return mvar'

syntax (name := trichotomy) "trichotomy" term "," term : tactic

@[tactic trichotomy] def evalTrichotomy : Tactic := fun stx =>
  withMainContext do
    trace[smt.profile.reconstruct] m!"[trichotomy] start time: {← IO.monoNanosNow}ns"
    let h₁ ← notLeToLt (← elabTerm stx[1] none)
    let h₂ ← notLeToLt (← elabTerm stx[3] none)
    let fname ← mkFreshId
    let mvar ← getMainGoal
    let mvar' ← trichotomyMeta mvar h₁ h₂ fname
    replaceMainGoal [mvar']
    evalTactic (← `(tactic| exact $(mkIdent fname)))
    trace[smt.profile.reconstruct] m!"[trichotomy] end time: {← IO.monoNanosNow}ns"
where
  notLeToLt : Expr → MetaM Expr := λ e => do
    match ← inferType e with
      | app (app (app (app (app (const ``LE.le ..) ..) ..) ..) ..) _ => mkAppM ``not_gt_of_le #[e]
      | app (app (app (app (const ``LE.le ..) ..) ..) ..) _ => mkAppM ``not_gt_of_le #[e]
      | app (app (app (app (app (const ``GE.ge ..) ..) ..) ..) ..) _ => mkAppM ``not_lt_of_ge #[e]
      | app (app (app (app (const ``GE.ge ..) ..) ..) ..) _ => mkAppM ``not_lt_of_ge #[e]
      | _ => pure e

end Smt.Reconstruct.Arith
