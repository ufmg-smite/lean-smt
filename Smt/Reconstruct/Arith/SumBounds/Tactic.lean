/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Tomaz Gomes Mascarenhas
-/

import Lean

import Smt.Reconstruct.Arith.MulPosNeg.Lemmas
import Smt.Reconstruct.Arith.SumBounds.Lemmas
import Smt.Reconstruct.Arith.SumBounds.Instances
import Smt.Reconstruct.Util

namespace Smt.Reconstruct.Arith

open Lean
open Meta Elab.Tactic Expr

theorem castEQ : ∀ {a b : Int}, a = b → Real.intCast.intCast a = Real.intCast.intCast b := by
  intros a b h
  rw [h]

def getCastRelThm (rel : Name) : Name :=
  match rel with
  | ``LT.lt => ``castLT
  | ``LE.le => ``castLE
  | ``Eq    => ``castEQ
  | _ => `unreachable

def combineBounds (mvar : MVarId) : Expr → Expr → MetaM Expr := fun h₁ h₂ =>
  mvar.withContext do
    let t₁ ← inferType h₁
    let rel₁ ← getOp t₁
    let t₂ ← inferType h₂
    let rel₂ ← getOp t₂
    let tp₁ ← getOpType t₁
    let tp₂ ← getOpType t₂
    let (h₁, h₂) ←
      match tp₁, tp₂ with
      | const `Int .., const `Real .. =>
        let thm := getCastRelThm rel₁
        pure (← mkAppM thm #[h₁], h₂)
      | const `Real .., const `Int .. =>
        let thm := getCastRelThm rel₂
        pure (h₁, ← mkAppM thm #[h₂])
      | _, _ => pure (h₁, h₂)
    let thmName : Name ←
      match rel₁, rel₂ with
      | ``LT.lt , ``LT.lt => pure ``sumBounds₁
      | ``LT.lt , ``LE.le => pure ``sumBounds₂
      | ``LT.lt , ``Eq    => pure ``sumBounds₃
      | ``LE.le , ``LT.lt => pure ``sumBounds₄
      | ``LE.le , ``LE.le => pure ``sumBounds₅
      | ``LE.le , ``Eq    => pure ``sumBounds₆
      | ``Eq    , ``LT.lt => pure ``sumBounds₇
      | ``Eq    , ``LE.le => pure ``sumBounds₈
      | ``Eq    , ``Eq    => pure ``sumBounds₉
      | _      , _      => throwError "[sumBounds] invalid operation"
    mkAppM thmName #[h₁, h₂]
where
  getOpType : Expr → MetaM Expr
  | app (Expr.const ``Not ..) e' => getOpType e'
  | app (app (app (app (app (Expr.const _ ..) tp) ..) ..) ..) .. => pure tp
  | app (app (app (app (Expr.const _ ..) tp) ..) ..) .. => pure tp
  | app (app (app (Expr.const _ ..) tp) ..) .. => pure tp
  | _ => throwError "[getOp] invalid parameter"

def traceSumBounds (r : Except Exception Unit) : MetaM MessageData :=
  return match r with
  | .ok _ => m!"{checkEmoji}"
  | _     => m!"{bombEmoji}"

def sumBounds (mv : MVarId) (hs : Array Expr) : MetaM Unit := withTraceNode `smt.reconstruct.sumBounds traceSumBounds do
  mv.withContext do
  if hs.isEmpty then
    throwError "[sum_bounds]: empty list of premisses"
  let h ← hs[1:].foldlM (combineBounds mv) hs[0]!
  mv.assign h

def sumBoundsMeta (mvar : MVarId) (h : Expr) (hs : List Expr) (name : Name)
    : MetaM MVarId :=
  mvar.withContext do
    go h hs
where
  go : Expr → List Expr → MetaM MVarId
    | acc, [] => do
      let type ← inferType acc
      let (_, mvar') ← MVarId.intro1P $ ← mvar.assert name type acc
      return mvar'
    | acc, h::hs => do
      let acc ← combineBounds mvar acc h
      go acc hs

namespace Tactic

syntax (name := sumBounds) "sum_bounds" "[" term,* "]" : tactic

def parseSumBounds : Syntax → TacticM (List Expr)
  | `(tactic| sum_bounds [$[$hs],*]) =>
    hs.toList.mapM (fun stx => elabTerm stx.raw none)
  | _ => throwError "[sum_bounds]: expects a list of premisses"

@[tactic sumBounds] def evalSumBounds : Tactic := fun stx =>
  withMainContext do
    trace[smt.debug] m!"[sum_bounds] start time: {← IO.monoNanosNow}ns"
    let (h, hs) ←
      match ← parseSumBounds stx with
      | h::hs =>
        let h'::hs' := List.reverse (h::hs) | throwError "unreachable"
        pure (h', hs')
      | [] => throwError "[sum_bounds]: empty list of premisses"
    let mvar ← getMainGoal
    let fname ← mkFreshId
    let mvar' ← sumBoundsMeta mvar h hs fname
    replaceMainGoal [mvar']
    evalTactic (← `(tactic| exact $(mkIdent fname)))
    trace[smt.debug] m!"[sum_bounds] end time: {← IO.monoNanosNow}ns"

end Smt.Reconstruct.Arith.Tactic
