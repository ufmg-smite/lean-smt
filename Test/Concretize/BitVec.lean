import Smt.Tactic.Concretize

def BitVec (w : Nat) := Fin (2^w)

protected def BitVec.zero (w : Nat) : BitVec w :=
  ⟨0, Nat.pos_pow_of_pos _ <| by decide⟩

instance : Inhabited (BitVec w) := ⟨BitVec.zero w⟩

opaque BitVec.xor {w : Nat} : BitVec w → BitVec w → BitVec w

def polyAdd {w : Nat} : BitVec w → BitVec w → BitVec w :=
  BitVec.xor
def polyDouble {w : Nat} (x : BitVec w) : BitVec w :=
  polyAdd x x

example (x y : BitVec 2) : polyDouble (polyAdd x y) = polyDouble (polyAdd y x) := by
  concretize [polyDouble, polyAdd]
  trace_state
  sorry
