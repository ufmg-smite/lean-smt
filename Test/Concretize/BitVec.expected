Test/Concretize/BitVec.lean:17:0: warning: declaration uses 'sorry'
x y : BitVec 2
polyAdd.«2» : BitVec 2 → BitVec 2 → BitVec 2 := BitVec.xor
polyDouble.«2» : BitVec 2 → BitVec 2 := fun x => polyAdd.«2» x x
⊢ polyDouble.«2» (polyAdd.«2» x y) = polyDouble.«2» (polyAdd.«2» y x)
